src/ripple/nodestore/DeterministicShard.md

```diff
@@ -0,0 +1,103 @@
+
+nudb.key:
+```
+char[8]         Type            The characters "nudb.key"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint64          Salt            A random seed
+uint64          Pepper          The salt hashed
+uint16          BlockSize       Size of a file block in bytes
+uint16          LoadFactor      Target fraction in 65536ths
+uint8[56]       Reserved        Zeroes
+uint8[]         Reserved        Zero-pad to block size
+```
+
+nudb.dat:
+```
+char[8]         Type            The characters "nudb.dat"
+uint16          Version         Holds the version number
+uint64          UID             Unique ID generated on creation
+uint64          Appnum          Application defined constant
+uint16          KeySize         Key size in bytes
+uint8[64]       (reserved)      Zeroes
+```
+there all fields are saved using network byte order (most significant byte first).
+
+To make the shard deterministic the following parameters are used as values of header field both for `nudb.key` and `nudb.dat` files.
+```
+Version         2
+UID             digest(0)
+Appnum          digest(2) | 0x5348524400000000 /* 'SHRD' */
+KeySize         32
+Salt            digest(1)
+Pepper          XXH64(Salt)
+BlockSize       0x1000 (4096 bytes)
+LoadFactor      0.5 (numeric 0x8000)
+```
+Note: XXH64() is well-known hash algorithm.
+
+The `digest(i)` mentioned above defined as the follows:
+
+First, RIPEMD160 hash `H` calculated of the following structure
+```
+uint256         lastHash        Hash of last ledger in shard
+uint32          index           Index of the shard
+uint32          firstSeq        Sequence number of first ledger in the shard
+uint32          lastSeq         Sequence number of last ledger in the shard
+uint32          version         Version of shard, 2 at the present
+```
+there all 32-bit integers are hashed in network byte order.
+
+Then, `digest(i)` is defined as the following portion of the above hash `H`:
+```
+digest(0) = H[0] << 56 | H[2] << 48 | ... | H[14] << 0,
+digest(1) = H[1] << 56 | H[3] << 48 | ... | H[15] << 0,
+digest(2) = H[19] << 24 | H[18] << 16 | ... | H[16] << 0,
+```
+where `H[i]` denotes `i`-th byte of hash `H`.
+
+
+## Contents
+
+After deterministic shard is created using the above mentioned headers, it filled with objects. First, all objects of the shard are collected and sorted in according to their hashes. Here the objects are: ledgers, SHAmap tree nodes including accounts and transactions, and final key object with hash 0. Objects are sorted by increasing of their hashes, precisely, by increasing of hex representations of hashes in lexicographic order. 
+
+For example, the following is an example of sorted hashes in their hex representation:
+```
+0000000000000000000000000000000000000000000000000000000000000000
```

Comment:
What value would be stored at that key?

Reply:
This is so called final key of the shard. Please refer to rippled code.

Reply:
@MarkusTeufelberger The value stored in the final key is the shard version, first ledger sequence, last ledger sequence and the hash of the ledger with the highest sequence (last ledger).

Reply:
Is it necessary to sort the node hashes? When a deterministic shard is written, we iterate through the node objects in a determined manner. The hash order isn't important, the sequence in which they are written is.

Reply:
As long as the order of insertion is clearly documented and re-implementable instead of being an emergent behavior of this code base, it can be any order... sorting just makes the order really easy and obvious.

Reply:
Done. Note: DeterministicShard class do not perform any sort, but NuDB do it when grab objects from hash table. As for now, ledgers divided into groups, objects are sorted within each group. See MD file for details.

Reply:
Urgh, this means I need to re-implement NuDB and node parsing (at least down to the level of header/leaf/inner node) before I can start generating deterministic shards if I don't want to write this in C++... before it was possible to just look at key/value pairs as binary blobs.

I also don't see where "natural SHAmap tree traversal order" is documented - https://github.com/ripple/rippled/tree/develop/src/ripple/shamap#walking-a-shamap talks of C++ constructs like pointers and seems geared towards finding/walking towards a certain leaf node instead of listing all nodes in some canonical representation.

Reply:
Please discuss Miguel Portilla directly. I do not understand what do you want.

Reply:
Documentation for that "natural SHAmap tree traversal order" - which MD file are you referring to?

Reply:
I don't know if such a documentation exists. Last fix to Deterministic shards MD which mentions "natural order" was made to reflect the code changes suggested by [this](https://github.com/miguelportilla/rippled/commit/62942f67848419e241911bdce39b28916947b91b) commit of Miguel Portilla. Prior to this change there was no need to mention "natural order". So I still recommend discuss this problem with him.

Reply:
@miguelportilla: can you please follow up with @MarkusTeufelberger. We should document the **precise** semantics so that the shard format can be parsed/generated by third-party implementations.

Reply:
I added description of the order in which objects are visited.